tests/utils.py

```diff
@@ -66,3 +72,111 @@ def get_node_name(nodename, ssh_config=None):
+
+    By default, this helper performs 3 retries with an exponential sleep
+    interval between each request and only retries internal server errors(500)
+    & service unavailable errors(503)
+
+    Arguments:
+        retries:          The number of retries to perform before giving up
+        backoff_factor:   The sleep interval between requests computed as
+                          {backoff factor} * (2 ^ ({number retries} - 1))
+        status_forcelist: HTTP status codes that we should force a retry on
+        method_whitelist: uppercased HTTP methods that we should retry
+        session:          Used to create a session
+
+    Returns:
+        A `requests.Session` object configured for retry.
+    """
+    session = session or requests.Session()
+    retry = Retry(
+        total=retries,
+        read=retries,
+        connect=retries,
+        backoff_factor=backoff_factor,
+        status_forcelist=status_forcelist,
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+    session.mount("https://", adapter)
+    return session
+
+
+def kubectl_exec(
+    host,
+    command,
+    pod,
+    kubeconfig='/etc/kubernetes/admin.conf',
+    **kwargs
+):
+    """Grab the return code from a `kubectl exec`"""
+    kube_args = ['--kubeconfig', kubeconfig]
+
+    if kwargs.get('container'):
+        kube_args.extend(['-c', kwargs.get('container')])
+    if kwargs.get('namespace'):
+        kube_args.extend(['-n', kwargs.get('namespace')])
+
+    kubectl_cmd_tplt = 'kubectl exec {} {} -- {}'
+
+    with host.sudo():
+        output = host.run(
+            kubectl_cmd_tplt.format(
+                pod,
+                ' '.join(kube_args),
+                ' '.join(command)
+            )
+        )
+        return output
+
+
+def run_salt_command(host, command, ssh_config):
+    """Run a command inside the salt-master container."""
+
+    pod = 'salt-master-{}'.format(
+        get_node_name('bootstrap', ssh_config)
+    )
+
+    output = kubectl_exec(
+        host,
+        command,
+        pod,
+        container='salt-master',
+        namespace='kube-system'
+    )
+
+    assert output.exit_status == 0, \
+        'command {} failed with: \nout: {}\nerr:'.format(
+            command,
+            output.stdout,
+            output.stderr
+        )
+
+
+def get_dict_element(data, path, delimiter='.'):
+    """
+    Traverse a dict using a 'delimiter' on a target string.
+    getitem(a, b) returns the value of a at index b
+    """
+    return functools.reduce(operator.getitem, path.split(delimiter), data)
+
+
+def set_dict_element(data, path, value, delimiter='.'):
+    """
+    Traverse a nested dict using a delimiter on a target string
+    replaces the value of a key within a dictionary and returns the new dict
+    """
+    path, _, key = path.rpartition(delimiter)
+    (get_dict_element(data, path) if path else data)[key] = value
+    return data
```

Comment:
Not used

Reply:
It is a cherry-pick from a commit that was only merged in 2.5
Atleast we use the `get_dict_element` method, so!